In [1]:
import numpy as np
import pandas as pd


In [2]:
movies = pd.read_csv("tmdb_5000_movies.csv")
credits = pd.read_csv("tmdb_5000_credits.csv")

In [4]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [8]:
movies.shape

(4803, 20)

In [5]:
credits.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [12]:
# merging on the basis of 'title' because it is not feasible to work on two different dataset. Since "title" column is common in both, we can merge them on the basis of 'title'
movies.merge(credits,on='title')

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4804     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4805       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4806          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4807          0                                                 []   
4808          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4804                                                NaN    9367   
4805                                                NaN   72766   
4806  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4807                        http://shanghaicalling.com/  126186   
4808                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...                es   
4805                                                 []                en   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4807                                                 []                en   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4804  El Mariachi just wants to play his guitar and ...   14.269792   
4805  A newlywed couple's honeymoon is upended by th...    0.642552   
4806  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4807  When ambitious New York attorney Sam is sent t...    0.857008   
4808  Ever since the second grade when he first saw ... 

In [11]:
 movies.merge(credits,on='title').shape

(4809, 23)

In [13]:
# we have to store this merged table in some variable and therefore we are merging and storing it in the 'movies' variable itself.
movies = movies.merge(credits,on='title')

In [14]:
movies.shape

(4809, 23)

Since we have 23 columns, we will remove those columns that are not necessary for our recommendation system. Below is the list of all those columns
which we are keeping
1. genres
2. id or movie_id
3. keywords
4. title       (we can also keep 'original_title' but we are not because original_title might have a title that is in a regional language)
5. overview

           (we can keep popularity column but since we are building a recommender system based on tags creation, this column will not fit in) 
7. cast        (because some people watch some movies just because of the actor) 
8. crew        (some people want to watch movies of a particular director, so crew is an important column)

In [15]:
movies[['movie_id','title','overview','genres','keywords','cast','crew']]
# storing these columns in 'movies' dataframe itself

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                                 genres  \
0     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...                                                 ...   
4804  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4805  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4806  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4807                                                 []   
4808                [{"id": 99, "name": "Documentary"}]   

                                               keywords  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4     [{"id": 818, "name": "based on novel"}, {"id":...   
...                                                 ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...   
4805                                                 []   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...   
4807                                                 []   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4804  [{"cast_id": 1, "character": "El Mariachi", "c...   
4805  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4806  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4807  [{"cast_id": 3, "character": "Sam", "credit_id...   
4808  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  
0     [{"credit_id": "52fe48009251416c750aca23", "de...  
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2     [{"credit_id": "54805967c3a36829b5002c41", "de...  
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
...                                                 ...  
4804  [{"credit_id": "52fe44eec3a36847f80b2

In [16]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [17]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [18]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

Now the next step is that we will create a new dataframe that will have a new column 'tags' created by merging 'overview', 'genres', 'keywords', 'cast', 'crew'

In [20]:
# dropping the null values row because these are just 3 rows and its not a large value. won't make much impact on the recommendation if not there
movies.dropna(inplace = True)

In [21]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [22]:
movies.duplicated().sum()

0

In [24]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

The output is in weird format. It is a list of dictionaries.
We want the above output in the given format:
['Action','Adventure','Fantasy','Science Fiction']
For this we will create a helper function

In [28]:
convert('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

TypeError: string indices must be integers, not 'str'

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

this is a string and that is why it is not working. So we first have to convert this string of list into list.
For this we have a module called ast in python which has a function literal_eval() which does this work.

In [27]:
import ast
ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [29]:
# therefore I am creating the above function again which implements literal_eval() function
def convert(obj):
    L = []               # creating a list
    for i in ast.literal_eval(obj):
        L.append(i['name'])         # appending the value to the key 'name' from the dictionary to the list
    return L

In [30]:
convert('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

['Action', 'Adventure', 'Fantasy', 'Science Fiction']

the above function is working fine

In [31]:
movies['genres'].apply(convert)       # apply the convert function to the genres column of movies dataframe to convert list of dictionaries into list

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [32]:
# storing the above result in the same genres column
movies['genres'] = movies['genres'].apply(convert)

In [34]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

#### now we'll do the same thing in the 'keywords' column

In [35]:
movies.iloc[0].keywords

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [36]:
movies['keywords'].apply(convert)  # apply the convert function to the keywords column of movies dataframe to convert list of dictionaries into list

0       [culture clash, future, space war, space colon...
1       [ocean, drug abuse, exotic island, east india ...
2       [spy, based on novel, secret agent, sequel, mi...
3       [dc comics, crime fighter, terrorist, secret i...
4       [based on novel, mars, medallion, space travel...
                              ...                        
4804    [united states–mexico barrier, legs, arms, pap...
4805                                                   []
4806    [date, love at first sight, narration, investi...
4807                                                   []
4808            [obsession, camcorder, crush, dream girl]
Name: keywords, Length: 4806, dtype: object

In [37]:
# storing the above result in the same keywords column
movies['keywords'] = movies['keywords'].apply(convert)

In [38]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [41]:
movies.iloc[0].cast         # or we can also use movies['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

###### here we will take top 3 actors and not all the cast

In [45]:
# {"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}
# in this "Jake Sully" is the name of character in the movie and "Sam Worthington" is the actual name of the actor
# therefore, we will take the the actual name. The same above function is repeated with some minor changes

def convert3(obj):
    L = []               # creating a list
    counter = 0            # to take maximum of 3 casts
    for i in ast.literal_eval(obj):
        if counter != 3:
            L.append(i['name'])         # appending the value to the key 'name' from the dictionary to the list
            counter+=1
        else:
            break
    return L

In [46]:
movies['cast'].apply(convert3)  # apply the convert function to the cast column of movies dataframe to get top 3 casts and to convert list of dictionaries into list

0        [Sam Worthington, Zoe Saldana, Sigourney Weaver]
1           [Johnny Depp, Orlando Bloom, Keira Knightley]
2            [Daniel Craig, Christoph Waltz, Léa Seydoux]
3            [Christian Bale, Michael Caine, Gary Oldman]
4          [Taylor Kitsch, Lynn Collins, Samantha Morton]
                              ...                        
4804    [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...
4805         [Edward Burns, Kerry Bishé, Marsha Dietlein]
4806           [Eric Mabius, Kristin Booth, Crystal Lowe]
4807            [Daniel Henney, Eliza Coupe, Bill Paxton]
4808    [Drew Barrymore, Brian Herzlinger, Corey Feldman]
Name: cast, Length: 4806, dtype: object

In [47]:
# storing the above result in the same cast column
movies['cast'] = movies['cast'].apply(convert3)

In [48]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [49]:
movies['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

For us only those dictionaries matters where the values of job is director. So we'll make a function for this

In [50]:
def fetch_director(obj):
    L = []               # creating a list
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            L.append(i['name'])         # appending the value to the key 'name' from the dictionary to the list
            break
    return L

In [52]:
movies['crew'].apply(fetch_director) # apply the convert function to the crew column of movies dataframe to fetch the directors and to convert list of dictionaries into list

0           [James Cameron]
1          [Gore Verbinski]
2              [Sam Mendes]
3       [Christopher Nolan]
4          [Andrew Stanton]
               ...         
4804     [Robert Rodriguez]
4805         [Edward Burns]
4806          [Scott Smith]
4807          [Daniel Hsia]
4808     [Brian Herzlinger]
Name: crew, Length: 4806, dtype: object

In [53]:
# storing the above result in the same crew column
movies['crew'] = movies['crew'].apply(fetch_director)

In [54]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   

                                               cast              crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]      [Sam Mendes]

In [55]:
# Now coming to the 'overview' column
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

Now this is a string. We will convert it into a list so that we can concatenate it with the other lists in the dataframe

In [56]:
movies['overview'].apply(lambda x:x.split())

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [57]:
# merging the above obtained lists into the same column
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [58]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   

                                               cast              crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]      [Sam Mendes]

Now since these columns  ('overview', 'genres', 'keywords', 'cast', 'crew') are a list, we will concatenate them to make a big list and store them in a column called 'tags'. Then we'll convert this list in a string to make it a paragraph.

We have to remove spaces from ('genres', 'keywords', 'cast', 'crew') columns. For that we have to apply transformation on these columns.
For exammple, we have to convert 'Sam Worthington' to 'SamWorthington'

In [59]:
# first we'll try to apply this on genres column
movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])

0       [Action, Adventure, Fantasy, ScienceFiction]
1                       [Adventure, Fantasy, Action]
2                         [Action, Adventure, Crime]
3                   [Action, Crime, Drama, Thriller]
4                [Action, Adventure, ScienceFiction]
                            ...                     
4804                       [Action, Crime, Thriller]
4805                               [Comedy, Romance]
4806               [Comedy, Drama, Romance, TVMovie]
4807                                              []
4808                                   [Documentary]
Name: genres, Length: 4806, dtype: object

In [61]:
# storing the above result in the same genres column
movies['genres'] = movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])

In [63]:
# now doing the same for remaining 3 columns
movies['keywords'] = movies['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movies['cast'] = movies['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movies['crew'] = movies['crew'].apply(lambda x:[i.replace(" ","") for i in x])

In [64]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   

                                            cast             crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]  [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]      [SamMendes]

In [65]:
# concatenating the above decided columns and forming a new column
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [66]:
movies.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   

                                            cast             crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]  [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]      [SamMendes]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...

In [67]:
new_df = movies[['movie_id','title','tags']]

In [68]:
new_df.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...

In [69]:
# now we'll convert the lists in tags column to string
new_df['tags'].apply(lambda x:" ".join(x))

0       In the 22nd century, a paraplegic Marine is di...
1       Captain Barbossa, long believed to be dead, ha...
2       A cryptic message from Bond’s past sends him o...
3       Following the death of District Attorney Harve...
4       John Carter is a war-weary, former military ca...
                              ...                        
4804    El Mariachi just wants to play his guitar and ...
4805    A newlywed couple's honeymoon is upended by th...
4806    "Signed, Sealed, Delivered" introduces a dedic...
4807    When ambitious New York attorney Sam is sent t...
4808    Ever since the second grade when he first saw ...
Name: tags, Length: 4806, dtype: object

In [70]:
# saving this in tags column
new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))

C:\Users\user\AppData\Local\Temp\ipykernel_6532\1942866522.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))


In [71]:
new_df.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...

In [72]:
# converting the string into lowercase
new_df['tags'].apply(lambda x:x.lower())

0       in the 22nd century, a paraplegic marine is di...
1       captain barbossa, long believed to be dead, ha...
2       a cryptic message from bond’s past sends him o...
3       following the death of district attorney harve...
4       john carter is a war-weary, former military ca...
                              ...                        
4804    el mariachi just wants to play his guitar and ...
4805    a newlywed couple's honeymoon is upended by th...
4806    "signed, sealed, delivered" introduces a dedic...
4807    when ambitious new york attorney sam is sent t...
4808    ever since the second grade when he first saw ...
Name: tags, Length: 4806, dtype: object

In [73]:
# saving the result in the same column
new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())

C:\Users\user\AppData\Local\Temp\ipykernel_6532\2284866200.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())


In [74]:
new_df.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...

# Vectorization
Here we will check the similarities between tags so that we could recommend similar movies.

For this, we will combine all the tags and create a large_text. For example, large_text = tag1 + tag2 + ..... +tag5000

Now, from this large_text, we will take out lets say 5000 most common words and create vectors with them.

We will use CountVectorizer from Sklearn.feature_extraction.text for this. Here we will use the stopwords functionality. Stopwords are the words which are used in the sentence formation. For example: are, is, in, the, etc.
We will neglect these stopwords

In [85]:
import nltk

In [86]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [87]:
def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
    return" ".join(y)

In [99]:
stem('loved loving love')

'love love love'

In [100]:
stem('dance dancing danced')

'danc danc danc'

In [89]:
new_df['tags'].apply(stem)  # performing stemming

0       in the 22nd century, a parapleg marin is dispa...
1       captain barbossa, long believ to be dead, ha c...
2       a cryptic messag from bond’ past send him on a...
3       follow the death of district attorney harvey d...
4       john carter is a war-weary, former militari ca...
                              ...                        
4804    el mariachi just want to play hi guitar and ca...
4805    a newlyw couple' honeymoon is upend by the arr...
4806    "signed, sealed, delivered" introduc a dedic q...
4807    when ambiti new york attorney sam is sent to s...
4808    ever sinc the second grade when he first saw h...
Name: tags, Length: 4806, dtype: object

In [90]:
# storing the above result in the same column
new_df['tags'] = new_df['tags'].apply(stem)

C:\Users\user\AppData\Local\Temp\ipykernel_6532\3827281523.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(stem)


now to remove words which have similar meaning
for example ['loved','loving','love']    =>  ['love','love','love']
we will use stemming for this

we will use nltk library for this.


we have created a stem function above for this thing.

In [91]:
from sklearn.feature_extraction.text import CountVectorizer

In [92]:
cv = CountVectorizer(max_features=5000, stop_words='english')    # max_features=5000 these are the words taken out from large_text

In [93]:
cv.fit_transform(new_df['tags']).toarray()      # explicitly converting tags to numpy array as cv returns a sparse matrix

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [94]:
# saving the above result in a new variable
vectors = cv.fit_transform(new_df['tags']).toarray()

In [95]:
vectors.shape

(4806, 5000)

In [96]:
# to check 5000 most frequent words
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      dtype=object)

##### now we have 4806 vectors which have 5000 numbers each
##### we need to calculate the distance between each and every movie to each and every other movie 
##### the more the distance, lesser the similarity between those 2 movies
##### here we will not calculate the euclean distance, instead we will calculate the cosine distance
##### i.e., euclean distance is sqrt((x2-x1)^2 - (y2-y1)^2) and cosine distance is the angle between the two vectors
### Note: If you are working with high dimensional data, as in now as we have 4806x5000, euclean distance fails. It is not a reliable measure for higher dimensional data.

##### for cosine similarity, we have a function in sklearn.metrics.pairwise known as cosine_similarity which ranges between 0 to 1. If it is 1 then similarity is more else not.

In [102]:
from sklearn.metrics.pairwise import cosine_similarity    

In [103]:
cosine_similarity(vectors)

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]])

In [104]:
cosine_similarity(vectors).shape

(4806, 4806)

In [125]:
# storing all these distances in a variable
similarity = cosine_similarity(vectors)

In [126]:
similarity[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

it has 4806 items as the distance is calculated between the 1st movie and all the other 4806 movies 
similarity[0].shape

In [127]:
# similarity score of 1st movie is 1 as it is the maximum similarity.. similarly it goes for 2nd movie
similarity[1]

array([0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
       0.02615329])

now the next step is to create a function which takes movie as an input and returns 5 similar movies.

To do this, first a movie is taken as input. Then we have to find the index of the movie in the data. Corresponding to that index, we will go to the similarity matrix which contains the distance of that movie with other movies. Then we will sort the similarity score in descending order i.e., more similar movies will come in the beginning. (more the similarity score, more similar the movie). Then we will fetch the top 5 movies and display them.

### this is how we will take out the index of the movie

In [128]:
new_df['title'] == 'Avatar'

0        True
1       False
2       False
3       False
4       False
        ...  
4804    False
4805    False
4806    False
4807    False
4808    False
Name: title, Length: 4806, dtype: bool

In [129]:
new_df[new_df['title'] == 'Avatar']

movie_id   title                                               tags
0     19995  Avatar  in the 22nd century, a parapleg marin is dispa...

In [130]:
new_df[new_df['title'] == 'Avatar'].index

Index([0], dtype='int64')

In [131]:
new_df[new_df['title'] == 'Avatar'].index[0]

0

In [135]:
new_df[new_df['title'] == 'Spectre'].index[0]

2

In [136]:
# this is how we'll fetch the similarity scores from similarity matrix.
similarity[new_df[new_df['title'] == 'Spectre'].index[0]]

array([0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
       0.        ])

In [144]:
# this is how we will sort the score
sorted(similarity[2],reverse = True)   # if we omit reverse attribute or write false in it then the sorting will be in ascending order

[0.9999999999999999,
 0.3609974361905776,
 0.3307189138830738,
 0.32639560491693337,
 0.2611164839335467,
 0.2390457218668787,
 0.22941573387056174,
 0.22645540682891913,
 0.22613350843332267,
 0.22592402852876595,
 0.220630652030457,
 0.21221867190679128,
 0.19094065395649332,
 0.18446619684315546,
 0.18042195912175807,
 0.17928429140015903,
 0.17817416127494956,
 0.17677669529663687,
 0.1735912687073534,
 0.171601618245911,
 0.17149858514250882,
 0.16984155512168936,
 0.16839382851364085,
 0.1670382761952652,
 0.16666666666666666,
 0.16413304107465318,
 0.16248342090492035,
 0.15811388300841897,
 0.15811388300841897,
 0.15811388300841894,
 0.1543033499620919,
 0.15386436372416593,
 0.15309310892394862,
 0.15231794896123557,
 0.15158476564770806,
 0.1507556722888818,
 0.1507556722888818,
 0.15006126199969522,
 0.1494035761667992,
 0.1494035761667992,
 0.14744195615489714,
 0.14731391274719738,
 0.147087101353638,
 0.147087101353638,
 0.14569287935358963,
 0.1453095484297741,
 0.144337

In [145]:
# we have to hold the positions of distance as after sorting, the indexes will change and we won't be able to find out which score at which index belongs to which movie
# for this we will use enumerate() function which returns an object which when converted to list given the following result
list(enumerate(similarity[2]))

[(0, 0.08603090020146065),
 (1, 0.06063390625908323),
 (2, 0.9999999999999999),
 (3, 0.060048057676907664),
 (4, 0.0773565934694095),
 (5, 0.07086833868923009),
 (6, 0.02192645048267573),
 (7, 0.1332504477222565),
 (8, 0.06454972243679027),
 (9, 0.10540925533894598),
 (10, 0.08385254915624211),
 (11, 0.3609974361905776),
 (12, 0.06565321642986127),
 (13, 0.049029033784546004),
 (14, 0.05590169943749474),
 (15, 0.06846531968814575),
 (16, 0.057353933467640436),
 (17, 0.15231794896123557),
 (18, 0.0827605888602368),
 (19, 0.0679018095318889),
 (20, 0.03162277660168379),
 (21, 0.059761430466719674),
 (22, 0.036084391824351615),
 (23, 0.1270001270001905),
 (24, 0.058123819371909635),
 (25, 0.0),
 (26, 0.11180339887498948),
 (27, 0.0679018095318889),
 (28, 0.09525009525014289),
 (29, 0.3307189138830738),
 (30, 0.07284643967679481),
 (31, 0.06933752452815364),
 (32, 0.04642383454426296),
 (33, 0.07071067811865475),
 (34, 0.0),
 (35, 0.07216878364870323),
 (36, 0.10783277320343841),
 (37, 0.0

In [148]:
# now we can sort the above data
sorted(list(enumerate(similarity[2])),reverse=True,key=lambda x:x[1])    # if key was not written, sorting would have been done on the basis of index values

[(2, 0.9999999999999999),
 (11, 0.3609974361905776),
 (29, 0.3307189138830738),
 (1344, 0.32639560491693337),
 (4075, 0.2611164839335467),
 (1745, 0.2390457218668787),
 (3338, 0.22941573387056174),
 (3164, 0.22645540682891913),
 (1719, 0.22613350843332267),
 (1132, 0.22592402852876595),
 (4343, 0.220630652030457),
 (147, 0.21221867190679128),
 (166, 0.19094065395649332),
 (3253, 0.18446619684315546),
 (3287, 0.18042195912175807),
 (227, 0.17928429140015903),
 (1101, 0.17817416127494956),
 (4641, 0.17677669529663687),
 (277, 0.1735912687073534),
 (1201, 0.171601618245911),
 (4008, 0.17149858514250882),
 (1715, 0.16984155512168936),
 (1085, 0.16839382851364085),
 (1839, 0.1670382761952652),
 (218, 0.16666666666666666),
 (972, 0.16413304107465318),
 (170, 0.16248342090492035),
 (108, 0.15811388300841897),
 (3353, 0.15811388300841897),
 (2923, 0.15811388300841894),
 (3973, 0.1543033499620919),
 (973, 0.15386436372416593),
 (4697, 0.15309310892394862),
 (17, 0.15231794896123557),
 (46, 0.15

In [149]:
# fetching top 5 movies
sorted(list(enumerate(similarity[2])),reverse=True,key=lambda x:x[1])[1:6]

[(11, 0.3609974361905776),
 (29, 0.3307189138830738),
 (1344, 0.32639560491693337),
 (4075, 0.2611164839335467),
 (1745, 0.2390457218668787)]

## Now to implement all these functionalities discussed above, we will create a function

In [156]:
# function which takes movie as an input and returns 5 similar movies
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]
    distances = similarity[movie_index]
    movie_list = sorted(list(enumerate(distances)),reverse=True,key=lambda x:x[1])[1:6]

    for i in movie_list:
        print(new_df.iloc[i[0]].title)     # new_df.iloc[i[0]].title will fetch 5 movies name from the data. i[0] will give the top 5 movies index and corresponding to that index we will fetch the movie name

In [157]:
recommend('Avatar')

Aliens vs Predator: Requiem
Aliens
Falcon Rising
Independence Day
Titan A.E.


In [160]:
recommend('The Avengers')

Iron Man 3
Avengers: Age of Ultron
Captain America: Civil War
Captain America: The First Avenger
Iron Man


In [161]:
recommend('Batman')

Batman
Batman & Robin
Batman Begins
Batman Returns
The R.M.


In [162]:
import pickle

In [165]:
pickle.dump(new_df.to_dict(),open('movies-recommender/movie_dict.pkl','wb'))

In [166]:
pickle.dump(similarity,open('movies-recommender/similarity.pkl','wb'))